lib/mix/lib/mix/tasks/test/stale.ex

```diff
@@ -0,0 +1,261 @@
+
+  def agent_write_manifest(pid) do
+    Agent.cast pid, fn sources ->
+      write_manifest(sources)
+      sources
+    end
+  end
+
+  @doc """
+  Should be called with the PID returned from `set_up`, after the ParallelRequire
+  is complete, whether it succeeded or not.
+  """
+  def agent_stop(nil),
+    do: :noop
+
+  def agent_stop(pid) do
+    Agent.stop(pid, :normal, :infinity)
+  end
+
+  ## Setup helpers
+
+  defp test_helper_stale?(test_paths) do
+    test_paths
+    |> Stream.map(&Path.join(&1, "test_helper.exs"))
+    |> Mix.Utils.stale?([manifest()])
+  end
+
+  defp mtimes(sources) do
+    Enum.reduce(sources, %{}, fn source(source: source, external: external), map ->
+      Enum.reduce([source | external], map, fn file, map ->
+        Map.put_new_lazy(map, file, fn -> Mix.Utils.last_modified(file) end)
+      end)
+    end)
+  end
+
+  defp update_stale_sources(sources, removed, changed) do
+    sources =
+      Enum.reject(sources, fn source(source: source) -> source in removed end)
+    sources =
+      Enum.reduce(changed, sources, &List.keystore(&2, &1, source(:source), source(source: &1)))
+    sources
+  end
+
+  ## Manifest
+
+  defp manifest, do: Path.join(Mix.Project.manifest_path, @stale_manifest)
+
+  defp read_manifest() do
+    try do
+      manifest() |> File.read!() |> :erlang.binary_to_term()
+    else
+      [@manifest_vsn | t] -> t
+      _ -> []
+    rescue
+      _ -> []
+    end
+  end
+
+  defp write_manifest([]) do
+    manifest()
+    |> File.rm()
+
+    :ok
+  end
+
+  defp write_manifest(sources) do
+    manifest = manifest()
+
+    manifest
+    |> Path.dirname()
+    |> File.mkdir_p!()
+
+    manifest_data =
+      [@manifest_vsn | sources]
+      |> :erlang.term_to_binary(compressed: 9)
+
+    File.write!(manifest, manifest_data)
+  end
+
+  ## Test changed dependency resolution
+
+  defp tests_with_changed_references(test_sources) do
+    test_manifest = manifest()
+    [elixir_manifest] = Mix.Tasks.Compile.Elixir.manifests()
+
+    if Mix.Utils.stale?([elixir_manifest], [test_manifest]) do
+      elixir_manifest_entries =
+        Mix.Compilers.Elixir.read_manifest(elixir_manifest)
+        |> Enum.group_by(&elem(&1, 0))
+
+      stale_modules =
+        for CE.module(module: module, beam: beam) <- elixir_manifest_entries.module,
+            Mix.Utils.stale?([beam], [test_manifest]),
+            do: module,
+            into: MapSet.new()
+
+      stale_modules = find_all_dependant_on(stale_modules, elixir_manifest_entries.source, elixir_manifest_entries.module)
+
+      for module <- stale_modules,
+          source(source: source, runtime_references: r, compile_references: c) <- test_sources,
+          module in r ++ c,
+          do: source,
+          into: MapSet.new()
+    else
+      MapSet.new()
+    end
+  end
+
+  defp find_all_dependant_on(modules, sources, all_modules, resolved \\ MapSet.new())
+
+  defp find_all_dependant_on(modules, _sources, _all_modules, modules),
```

Comment:
This is a potentially expensive check because you need to traverse both modules to see if they are equal. I would use MapSet.size(new_modules) to see if they changed in the clause below and act accordingly.